src/query/functions/src/srfs/mod.rs

```diff
@@ -12,408 +12,117 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use common_expression::types::array::ArrayColumnBuilder;
-use common_expression::types::AnyType;
+use std::sync::Arc;
+
+use common_expression::types::nullable::NullableColumn;
 use common_expression::types::DataType;
 use common_expression::Column;
-use common_expression::ColumnBuilder;
 use common_expression::Function;
 use common_expression::FunctionEval;
 use common_expression::FunctionKind;
 use common_expression::FunctionProperty;
 use common_expression::FunctionRegistry;
 use common_expression::FunctionSignature;
 use common_expression::Scalar;
+use common_expression::ScalarRef;
 use common_expression::Value;
-use common_expression::ValueRef;
 
 pub fn register(registry: &mut FunctionRegistry) {
     registry.properties.insert(
         "unnest".to_string(),
         FunctionProperty::default().kind(FunctionKind::SRF),
     );
 
-    let unnest_impl = |args: &[ValueRef<AnyType>], num_rows| {
-        debug_assert_eq!(args.len(), 1);
-        let unnest_array = match args[0].clone().to_owned() {
-            Value::Scalar(Scalar::Array(col)) => {
-                ArrayColumnBuilder::<AnyType>::repeat(&col, num_rows).build()
+    registry.register_function_factory("unnest", |_, arg_types: &[DataType]| {
+        match arg_types {
+            [
+                ty @ (DataType::Null
+                | DataType::EmptyArray
+                | DataType::Nullable(_)
+                | DataType::Array(_)),
+            ] => Some(build_unnest(ty, Box::new(|ty| ty))),
+            _ => {
+                // Generate a fake function with signature `unnset(Array(T0 NULL))` to have a better error message.
+                Some(build_unnest(
+                    &DataType::Array(Box::new(DataType::Boolean)),
+                    Box::new(|ty| ty),
```

Comment:
It seems that the paramenter does not need to be a function, can pass `DataType` directly.